README.md
# jQuery Validate

jQuery Validate is a live input validation plugin.

### Supported Validation Types:
- Phone
- Numeric
- ABA Bank Routing
- Email Addresses
- Matching Other Inputs
- Passwords
- Required
- Date
- String
- Minimum Length
- Maximum Length

## Basic Usage

#### HTML
```html
<input type="text" id="first-name" />
```

#### JavaScript
```javascript
// Required Field
$('#first-name').validate( { required: true } );
```

By default, the plugin will add CSS classes to the validated input. You can
style these with CSS.

#### CSS
```css
.validate-success {
    border: 1px solid green;
}

.validate-error {
    border: 1px solid red;
}
```

## Options
| Type              | Parameter             |
| ---               | ---                   |
| Phone             | type: "phone"         |
| Email             | type: "email"         |
| Password          | type: "password"      |
| Numeric           | type: "number"        |
| Date              | type: "date"          |
| ABA Routing       | type: "ABA"           |
| Match             | match: "#other-input" |
| Required          | required: true        |
| Minimum Length    | minLength: 4          |
| Maximum Length    | maxLength: 8          |

## Callback

If you need to do something programmatically when the validation is done, onDone
will return the result as a boolean.

#### JavaScript
### onDone: function( result )
```javascript
// JavaScript anonymous callback function after each validation check (keyup).
$('#first-name').validate( {
    required: true,
    onDone: function( result ){
        if( !result ){
            console.log("Cannot leave input blank.");
        } else {
            console.log("You typed: "+$(this).val()+".");
        }
    }
});
``` 

CHANGELOG.md
#Change Log

## [Unreleased]

	- Cleaned up code and fixed indentation.

## [1.0.1] 2015-10-22
### Added

    - Default validation type of "string".

    - Minimum length and maximum length to validation options.
	
	- Base CSS class to signify that the input is being validated.
	
	- CSS classes for failed and successful validations.
	
### Changed

    - Removed old return statements in favor of onDone callback.

    - Replaced variables and cleaned up code.
	
## [1.0.0]
### Added
	
	- Input validations.
	
	- Created jQuery plugin scaffolding.

js/jquery-validate-min.js
(function(a){a.fn.validate=function(b){var d="jQuery Validate: ";console.log("hello");var c=a.extend({type:"string",passwordOptions:{uppercase:false,lowercase:false,number:false,special:false,minimum:8,maximum:null},onDone:null,required:false,minLength:0,maxLength:null,match:null,debug:false,force:false},b);return this.each(function(f,g){var e;a(g).addClass("validate");var h=function(){var o;o=a(this).val();if(o!==""||o.length!==0){if(c.type==="phone"){regex=new RegExp(/^(\()?[2-9]\d{2}(\))?(\s)*(-|)?(\s)*[2-9]\d{2}(\s)*(-|)?(\s)*\d{4}$/);if(regex.test(o)){if(c.debug){console.log(d+o+" is a phone number.")}e=true}else{if(c.debug){console.log(d+o+" is not a valid phone number.")}e=false}}else{if(c.type==="email"){regex=new RegExp(/^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);if(regex.test(o)){if(c.debug){console.log(d+o+" is a valid email address.")}e=true}else{if(c.debug){console.log(d+o+" is not a valid email address.")}e=false}}else{if(c.type==="password"){regexString="(^";if(c.passwordOptions.number){regexString+="(?=.*\\d)"}if(c.passwordOptions.lowercase){regexString+="(?=.*[a-z])"}if(c.passwordOptions.uppercase){regexString+="(?=.*[A-Z])"}if(c.passwordOptions.special){regexString+="(?=.*[\\W])"}regexString+=".{"+c.passwordOptions.minimum+",";if(c.passwordOptions.maximum>c.passwordOptions.minimum){regexString+=c.passwordOptions.maximum}regexString+="}$)";if(c.debug){console.log(d+regexString)}regex=new RegExp(regexString);if(regex.test(o)){if(c.debug){console.log(d+o+" is a valid password.")}e=true}else{if(c.debug){console.log(d+o+" is not a valid password.")}e=false}}else{if(c.type==="date"){var p,j,l,k;k=o.split("/");if(c.debug){console.log(d+"Date ("+o+")")}p=parseInt(k[0]);j=parseInt(k[1]);l=parseInt(k[2]);if(p>=1&&p<=12){if(j>=1&&j<=31){if(l>=1&&l<=new Date().getFullYear()){if((p===4||p===6||p===9||p===11)&&j===31){e=false}else{if(p===2){var m=(l%4===0&&(l%100!==0||l%400===0));if(j>29||(j===29&&!m)){e=false}else{e=true}}else{e=true}}}}}}else{if(c.type==="ABA"){n=0;for(i=0;i<o.length;i+=3){n+=parseInt(o.charAt(i),10)*3+parseInt(o.charAt(i+1),10)*7+parseInt(o.charAt(i+2),10)}if(n!==0&&n%10===0){e=true}else{e=false}}else{if(c.type==="number"){e=new RegExp(/^\d+$/).test(o)}else{if(c.type==="string"){if(typeof o==="string"){e=true}else{e=false}}else{if(c.debug){console.warn(d+"Validation type not recognized.")}}}}}}}}}else{e=false}if(c.required){if(o===""||o===0){e=false}else{e=true}}if(c.maxLength!==null){if(o.length>c.maxLength){e=false}}if(c.minLength!==null){if(o.length<c.minLength){e=false}}if(c.match){if(c.match instanceof jQuery){if(c.match.val()!==""){if(o===c.match.val()){e=true}else{e=false}}}else{if(typeof c.match==="string"){console.log("String");if(c.match!==""){if(a(c.match).val()!==""){if(o===a(c.match).val()){e=true}else{e=false}}}else{e=false}}}}if(!e){a(g).addClass("validate-error");a(g).removeClass("validate-success")}else{a(g).removeClass("validate-error");a(g).addClass("validate-success")}if(typeof c.onDone==="function"){c.onDone.call(this,e);return this}};a(g).on("keyup",function(){h.call(this)});if(c.force){h.call(this)}})}})(jQuery);

js/jquery.validate.js
/***************************
   jQuery Validate Plugin
   Author: Kyle Bedell
   Version: 1.0.1
****************************/
(function($){
	$.fn.validate = function( options ){
		var PLUGIN_TAG = 'jQuery Validate: ';

		var settings = $.extend({
			type: "string",
			passwordOptions: {
				uppercase: false,
				lowercase: false,
				number: false,
				special: false,
				minimum: 8,
				maximum: null
			},
			onDone: null,
			required: false,
                        minLength: 0,
                        maxLength: null,
			match: null,
			debug: false,
			force: false
		}, options );
		
		return this.each( function( idx, el ){
			
			var result;
                        
			// Add a reference class.
			$(el).addClass('validate');
                        
			var init = function(){
				var value;
                                
				// Get the value of the input.
				value = $(this).val();
                                
				// If the value is not blank.
				if( value !== "" || value.length !== 0 ){
					if( settings.type === "phone" ){
						regex = new RegExp(/^(\()?[2-9]\d{2}(\))?(\s)*(-|)?(\s)*[2-9]\d{2}(\s)*(-|)?(\s)*\d{4}$/);
						if( regex.test( value ) ){
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+' is a phone number.' );
							}
							result = true;
						} else {
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+" is not a valid phone number.");
							}
							result = false;
						}
					} else if ( settings.type === "email" ){
						regex = new RegExp(/^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
						if( regex.test( value ) ){
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+' is a valid email address.' );
							}
							result = true;
						} else {
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+" is not a valid email address.");
							}
							result = false;
						}
					} else if ( settings.type === "password" ){
						// Start the regular expression.
						regexString = "(^";
						
						// Require a number?
						if( settings.passwordOptions.number ){
							regexString += "(?=.*\\d)";
						}
						
						// Require a lower case letter?
						if( settings.passwordOptions.lowercase ){
							regexString += "(?=.*[a-z])";
						}
						
						// Require an upper case letter?
						if( settings.passwordOptions.uppercase ){
							regexString += "(?=.*[A-Z])";
						}
						
						if( settings.passwordOptions.special ){
							regexString += "(?=.*[\\W])";
						}
						
						// Minimum length. ( Default: 8 )
						regexString += ".{"+settings.passwordOptions.minimum+",";
						
						// Set a maximum length.
						if( settings.passwordOptions.maximum > settings.passwordOptions.minimum ){
							regexString += settings.passwordOptions.maximum;
						}
						
						// Close the regular expression.
						regexString += "}$)";
						
						// Log regular expression in debug mode.
						if( settings.debug ){
							console.log( PLUGIN_TAG+regexString );
						}
						
						// RegExp initialization.
						regex = new RegExp( regexString );
						
						// Test the string.
						if( regex.test( value ) ){
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+' is a valid password.' );
							}
							result = true;
						} else {
							if( settings.debug ){
								console.log( PLUGIN_TAG+value+" is not a valid password.");
							}
							result = false;
						}
					
					} else if( settings.type === "date" ){
						
						var month,
							day,
							year,
							date;
							
						date = value.split('/');
						
						if( settings.debug ){
							console.log( PLUGIN_TAG+'Date ('+value+')' );
						}
						
						month = parseInt( date[0] );
						day = parseInt( date[1] );
						year = parseInt( date[2] );
						
						if( month >= 1 && month <= 12 ){
							if( day >= 1 && day <= 31 ){
								if( year >= 1 && year <= new Date().getFullYear() ){
									if ( (month===4 || month===6 || month===9 || month===11) && day === 31 ){
										result = false;
									} else if ( month === 2 ){
										var isleap = (year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0));
										if (day > 29 || (day ===29 && !isleap)){
											result = false;
										} else {
											result = true;
										}
									} else {
										result = true;
									}
								}
							}
						}			
					} else if( settings.type === "ABA" ){
						// Run through each digit and calculate the total.
						n = 0;
						for (i = 0; i < value.length; i += 3) {
								n += parseInt( value.charAt(i), 10 ) * 3
								+  parseInt( value.charAt(i + 1), 10) * 7
								+  parseInt( value.charAt(i + 2), 10);
						}

						// If the resulting sum is an even multiple of ten (but not zero),
						// the aba routing number is good.
						if (n !== 0 && n % 10 === 0){
								result = true;
						} else {
								result = false;
						}
					} else if( settings.type === "number" ){
						result = new RegExp( /^\d+$/ ).test( value );
					} else if( settings.type === "string" ){
						if( typeof value === "string" ){
							result = true;
						} else {
							result = false;
						}
                    } else {
						if( settings.debug ){
							console.warn(PLUGIN_TAG+"Validation type not recognized.");
						}
					}
				} else {
					result = false;
				}
				
				// Is this a required field?
				if( settings.required ){
					// Perform an empty string check.
					if( value === "" || value === 0 ){
						result = false;
					} else {
						result = true;
					}
				}
                                
                                // Does the value exceed maximum length?
                                if( settings.maxLength !== null ){
                                        if( value.length > settings.maxLength ){
                                            result = false;
                                        }
                                }
                                
                                // Is the string shorter than the minimum length? 
				if( settings.minLength !== null ){
                                        if( value.length < settings.minLength ){
                                            result = false;
                                        }
                                }
                                
				// If the value needs to match a second input's value.
				if( settings.match ){
					
					// If the parameter is a jQuery object...
					if( settings.match instanceof jQuery ){
						// Check the value for a match.
						if( settings.match.val() !== "" ){
							if( value === settings.match.val() ){
								result = true;
							} else {
								result = false;
							}
						}
					// Else, if we have a CSS selector string...
					} else if( typeof settings.match === "string" ){
						console.log( "String" );
						// If the CSS selector string is not empty.
						if( settings.match !== "" ){
							if( $( settings.match ).val() !== "" ){
								// Get the jQuery object of the CSS selector string and test.
								if( value === $( settings.match ).val() ){
									result = true;
								} else {
									result = false;
								}
							}
						} else {
							result = false;
						}
					}
				}
                                
				// Update CSS classes.
				if( !result ){
					// Add error class.
					$( el ).addClass('validate-error');
					// Remove success class.
					$( el ).removeClass('validate-success');
				} else {
					// Add success class.
					$( el ).removeClass('validate-error');
					// Remove error class.
					$( el ).addClass('validate-success');
				}
				
				// Callback when done.
				if( typeof settings.onDone === "function" ){
					settings.onDone.call( this, result );
					// Return the jQuery object for chaining.
					return this;
				}
			};
			
			// Add keyup tests.
			$( el ).on('keyup', function(){
				init.call( this );
			});
			
			if( settings.force ){
				// Run initial test.
				init.call( this );
			}	
		});
	};
})(jQuery);